crates/holochain/src/core/workflow/app_validation_workflow.rs

```diff
@@ -658,70 +643,75 @@ where
+                            tracing::warn!("error fetching dependent hash {hash:?}: {err}");
+                        } else {
+                            // dependency has been fetched and added to the cache
+                            fetched_dependencies.lock().remove(&hash);
+                        }
+                    }
                 }
-                run_validation_callback_inner(
-                    invocation,
-                    ribosome,
-                    workspace_read,
-                    network,
-                    (fetched_deps, recursion_depth + 1),
-                    visited_activity,
-                )
-                .await
-            }
+            });
+            // await all fetches in a separate task without awaiting the task
+            // to finish
+            tokio::spawn(async { futures::future::join_all(fetches).await });
+            Ok(Outcome::AwaitingDeps(hashes))
         }
         ValidateResult::UnresolvedDependencies(UnresolvedDependencies::AgentActivity(
             author,
             filter,
         )) => {
-            if recursion_depth > 20 || visited_activity.contains(&filter) {
-                Ok(Outcome::AwaitingDeps(vec![author.into()]))
-            } else {
+            tracing::debug!(
+                "unresolved dependencies: agent activity of {author:?} with filter {filter:?}"
+            );
+            // fetch missing agent activities in the background without awaiting them
+            tokio::spawn({
                 let cascade_workspace = workspace_read.clone();
+                let author = author.clone();
                 let cascade =
                     CascadeImpl::from_workspace_and_network(&cascade_workspace, network.clone());
-                cascade
-                    .must_get_agent_activity(author.clone(), filter.clone())
-                    .await?;
-                visited_activity.insert(filter);
-                run_validation_callback_inner(
-                    invocation,
-                    ribosome,
-                    workspace_read,
-                    network,
-                    (fetched_deps, recursion_depth + 1),
-                    visited_activity,
-                )
-                .await
-            }
+
+                // keep track of which dependencies are being fetched to
+                // prevent multiple fetches of the same hash
+                let fetched_dependencies = fetched_dependencies.clone();
+                async move {
+                    let new_dependency = fetched_dependencies.lock().insert(author.clone().into());
+                    // fetch dependency if it is not being fetched yet
+                    if new_dependency {
+                        let result = cascade
+                            .must_get_agent_activity(author.clone(), filter)
+                            .await;
+                        if let Err(err) = result {
+                            tracing::warn!(
```

Comment:
Same question here about error handling, do we retry or reject the parent?

Reply:
see above